functions/Remove-DbaBackup.ps1

```diff
@@ -99,143 +99,158 @@ It will also remove any backup folders that no longer contain backup files.
+				throw "RetentionPeriod '$UserFriendlyRetention' format invalid! See Get-Help for correct formatting and examples."
+			}
+
+			switch ($Units)
+			{
+				'h' { $UnitString = 'Hours'; [datetime]$ReturnDatetime = (Get-Date).AddHours(-$Value)  }
+				'd' { $UnitString = 'Days';  [datetime]$ReturnDatetime = (Get-Date).AddDays(-$Value)   }
+				'w' { $UnitString = 'Weeks'; [datetime]$ReturnDatetime = (Get-Date).AddDays(-$Value*7) }
+				'm' { $UnitString = 'Months';[datetime]$ReturnDatetime = (Get-Date).AddMonths(-$Value) }
+			}
+			$ReturnDatetime
+		}
+
+		# Validations
+		# Ensure BackupFileExtension does not begin with a .
+		if ($BackupFileExtension -match "^[.]") {
+			Write-Message -Message "Parameter -BackupFileExtension begins with a period '$BackupFileExtension'. A period is automatically prepended to -BackupFileExtension and need not be passed in." -Warning
+		}
+
+	}
+	PROCESS
+	{
+		# Process stuff
+		Write-Message -Message "Started" -Level 3 -Silent $Silent
+		Write-Message -Message "Removing backups from $Path" -Level 3 -Silent $Silent
+		# Convert Retention Value to an actual DateTime
+		try {
+			$RetentionDate = Convert-UserFriendlyRetentionToDatetime -UserFriendlyRetention $RetentionPeriod
+			Write-Message -Message "Backup Retention Date set to $RetentionDate" -Level 5 -Silent $Silent
+		} catch {
+			throw $_
+		}
+
+		# Filter out unarchived files if -CheckArchiveBit parameter is used
+		if ($CheckArchiveBit.IsPresent) {
+			Write-Message -Message "Removing only archived files" -Level 5 -Silent $Silent
+			Filter DbaArchiveBitFilter {
+				If ($_.Attributes -notmatch "Archive") {
+					$_
+				}
+			}
+		} else {
+			Filter DbaArchiveBitFilter {
+				$_
+			}
+		}
+		# Enumeration may take a while. Without resorting to "esoteric" file listing facilities
+		# and given we need to fetch at least the LastWriteTime, let's just use "streaming" processing
+		# here to avoid issues like described in #970
+		Get-ChildItem $Path -Filter "*.$BackupFileExtension" -File -Recurse -ErrorAction SilentlyContinue -ErrorVariable EnumErrors | `
+			Where-Object LastWriteTime -lt $RetentionDate | DbaArchiveBitFilter | Foreach-Object `
+		{
+			$file = $_
+			if ($PSCmdlet.ShouldProcess($file.Directory.FullName, "Removing backup file $($file.Name)")) {
+				try {
+					$file
+					$file | Remove-Item -Force -EA Stop
+				} catch {
+					Write-Message -Message "Failed to remove $file" -Level Warning -ErrorRecord $_
+				}
+			}
+		}
+		if ($EnumErrors) {
+			Write-Message "Errors encountered enumerating files" -Level Warning -ErrorRecord $EnumErrors
+		}
+		Write-Message -Message "File Cleaning ended" -Level 3 -Silent $Silent
+		# Cleanup empty backup folders.
+		if ($RemoveEmptyBackupFolder.IsPresent) {
+			Write-Message -Message "Removing empty folders" -Level 3 -Silent $Silent
+			Get-ChildItem -Directory -Path $Path -Recurse -ErrorAction SilentlyContinue -ErrorVariable EnumErrors | Foreach-Object { $_.FullName } | Sort-Object -Descending | `
```

Comment:
Don't use `Foreach-Object { $_.FullName }` to extract a single property from an object
`Select-Object -ExpandProperty FullName` should be significantly faster.
Since it's the first command in the pipeline, you can even use regular braces for the job for even better performance:
`(Get-ChildItem -Directory -Path $Path -Recurse -ErrorAction SilentlyContinue -ErrorVariable EnumErrors).FullName`

Also: Please no backticks